analyses/cms-open-data-ttbar/ml.py

```diff
@@ -0,0 +1,276 @@
+        .Define(f"{feature_names[10]}", "Take(Jet_pt_cut, bH_idx)")
+        .Define(f"{feature_names[11]}", "Take(Jet_pt_cut, bL_idx)")
+        .Define(f"{feature_names[12]}", "Take(Jet_btagCSVV2_cut, W1_idx)")
+        .Define(f"{feature_names[13]}", "Take(Jet_btagCSVV2_cut, W2_idx)")
+        .Define(f"{feature_names[14]}", "Take(Jet_btagCSVV2_cut, bH_idx)")
+        .Define(f"{feature_names[15]}", "Take(Jet_btagCSVV2_cut, bL_idx)")
+        .Define("Jet_qgl_cut", "Jet_qgl[Jet_mask]")
+        .Define(f"{feature_names[16]}", "Take(Jet_qgl_cut, W1_idx)")
+        .Define(f"{feature_names[17]}", "Take(Jet_qgl_cut, W2_idx)")
+        .Define(f"{feature_names[18]}", "Take(Jet_qgl_cut, bH_idx)")
+        .Define(f"{feature_names[19]}", "Take(Jet_qgl_cut, bL_idx)")
+        #        jets' four-momenta
+        .Define(
+            "JetW1_p4",
+            f"ConstructP4({feature_names[8]}, JetW1_eta, JetW1_phi, Take(Jet_mass_cut, W1_idx))",
+        )
+        .Define(
+            "JetW2_p4",
+            f"ConstructP4({feature_names[9]}, JetW2_eta, JetW2_phi, Take(Jet_mass_cut, W2_idx))",
+        )
+        .Define(
+            "JetbL_p4",
+            f"ConstructP4({feature_names[11]}, JetbL_eta, JetbL_phi, Take(Jet_mass_cut, bL_idx))",
+        )
+        .Define(
+            "JetbH_p4",
+            f"ConstructP4({feature_names[10]}, JetbH_eta, JetbH_phi, Take(Jet_mass_cut, bH_idx))",
+        )
+    )
+
+    # # build features 8 other features
+    df = (
+        df.Define(
+            f"{feature_names[0]}",
+            "sqrt(pow(Lepton_eta.at(0)-JetbL_eta,2.)+pow(Lepton_phi.at(0)-JetbL_phi,2.))",
+        )
+        .Define(
+            f"{feature_names[1]}", "sqrt(pow(JetW1_eta-JetW2_eta,2.)+pow(JetW1_phi-JetW2_phi,2.))"
+        )
+        .Define(
+            f"{feature_names[2]}", "sqrt(pow(JetW1_eta-JetbH_eta,2.)+pow(JetW1_phi-JetbH_phi,2.))"
+        )
+        .Define(
+            f"{feature_names[3]}", "sqrt(pow(JetW2_eta-JetbH_eta,2.)+pow(JetW2_phi-JetbH_phi,2.))"
+        )
+        .Define(
+            f"{feature_names[4]}",
+            "return Map(JetbL_p4+Lepton_p4.at(0), [] (const ROOT::Math::PxPyPzMVector &p) {return p.M();})",
+        )
+        .Define(
+            f"{feature_names[5]}",
+            "return Map(JetW1_p4+JetW2_p4, [] (const ROOT::Math::PxPyPzMVector &p) {return p.M();})",
+        )
+        .Define(
+            f"{feature_names[6]}",
+            "return Map(JetW1_p4+JetW2_p4+JetbH_p4, [] (const ROOT::Math::PxPyPzMVector &p) {return p.M();})",
+        )
+        .Define(
+            f"{feature_names[7]}",
+            "return Map(JetW1_p4+JetW2_p4+JetbH_p4, [] (const ROOT::Math::PxPyPzMVector &p) {return p.Pt();})",
+        )
+    )
+    # put all features into one collection
+    feature_names = feature_names.__str__().replace("[", "{").replace("]", "}").replace("'", "")
+    df = df.Define("features", f"ROOT::VecOps::RVec<ROOT::RVecF>({feature_names})")
```

Comment:
do we ever use the columns defined above (containing the features) directly, or do we always access them via the `features` column?

if we never use them outside of the vector, it's a waste to go through all those `Define`s: there are copies involved, and they make the computation graph much larger. Could all of the above just be a single C++ function in `ml_helpers.cpp` called `eval_features` that returns the `RVec<RVecF>`?